### PR TITLE
Do not remove identities into Add and Mul when evaluate=False

### PR DESCRIPTION
**Pull request by cgabard**
As discussed on ML ( https://groups.google.com/forum/#!topic/sympy/kx-sZvDy2ac ), this change prevent `Mul` and `Add` from removing their identity element if evaluate=False.

From an external point of view, the behavior is now : 

```python
a = Add(0, 3, 2, evaluate=False)
assert a.func is Add
assert a.args == (0, 3, 2)
# Previously, it was:
#  assert a.args == (3, 2)
     
b = Mul(1, 3, 2, evaluate=False)
assert b.func is Mul
assert b.args == (1, 3, 2)
# Previously, it was:
# assert b.args == (3, 2)
```

All other changes are there to ensure backward compatibility in functions relying on the old behavior.

### CODE REVIEW

**Comment by asmeurer** [comment]
This looks good. With this we have

```
>>> with evaluate(False):
...     print(srepr((S(1)/3).args))
(Integer(1), Pow(Integer(3), Integer(-1)))
```

Should we change it to not include the 1? It doesn't seem to affect printing.

**Comment by cgabard** [comment]
Personally, I have no preference on this subject. Should I manage it in the div operator directly?

**Comment by asmeurer** [comment]
I think it's because One uses the `Rational.__div__`, which does `Rational(self.p*other.q, self.q*other.p, igcd(self.p, other.p)*igcd(self.q, other.q))` when evaluate=False. We could define `One.__div__`, which would be more efficient anyway, as it wouldn't do the unnecessary multiplications and gcd calculations.

**Comment by cgabard** [comment]
Hum, I tried but have some problems. I added a test for this in `test_add` of `test_evaluate.py`. It work when I run only this test but fail when run under complete test suite.

File: sympy/core/tests/test_evaluate.py [hunk 1]
```diff
@@ -44,22 +45,23 @@ def test_add():
...
-        assert S(6) / 3 == Mul(6, S(1) / 3)
+        assert S(6) / 3 == Mul(6, Pow(3, -1))
         assert S(1) / 3 * 6 == Mul(S(1) / 3, 6)
 
         assert S(1) / 3 + sqrt(3) == Add(S(1) / 3, sqrt(3))
         assert sqrt(3) + S(1) / 3 == Add(sqrt(3), S(1) / 3)
 
         assert S(1) / 2 * 10.333 == Mul(S(1) / 2, 10.333)
-        assert 10.333 * S(1) / 2 == Mul(10.333, S(1) / 2)
+        assert 10.333 * (S(1) / 2) == Mul(10.333, (S(1) / 2))
```
Comment (smichr) [inline sympy/core/tests/test_evaluate.py]: or `assert 10.333 * (S(1) / 2) == Mul(10.333, 2**-1)` ? or simply no parens on the right. just a nit.

**Comment by smichr** [comment]
> Should we change it to not include the 1? It doesn't seem to affect printing.

I am in favor of making `1/x` just be `Pow(x, -1)` rather than `1*Pow(x, -1)`

**Comment by asmeurer** [comment]
What is the failure? It could be a problem with caching.

**Comment by gxyd** [comment]
@cgabard can you please resolve the merge conflicts here?

**Comment by czgdp1807** [comment]
@cgabard Are you still working on it? Please resolve the merge conflicts. Thanks for the contribution.

**Comment by cgabard** [comment]
No sorry and I will not any time soon to do it.

Le dim. 22 sept. 2019 à 20:32, Gagandeep Singh <notifications@github.com> a
écrit :

**Comment by czgdp1807** [comment]
No problem, please let me know whenever you feel that your patch is ready for the review. Thanks.

**Comment by cgabard** [comment]
I moved from the company that need it in the first place, near two years
ago, and I forgot how we solved this before I am leaving. I am not sure I
will have any time to work on it on my free time soon.
I think you should find someone else to do the work or decline it because I
will not be able to do anything on it in near future.

Le mer. 2 oct. 2019 à 08:20, Gagandeep Singh <notifications@github.com> a
écrit :

**Comment by czgdp1807** [comment]
Ah! Sorry, I misread your first message. Thanks for the details and the contribution you made. :-) 

PS - Please do not close this PR.

**Comment by czgdp1807** [comment]
@asmeurer Can we continue this PR?

**Comment by czgdp1807** [comment]
ping @asmeurer

**Comment by czgdp1807** [comment]
@oscarbenjamin @jksuom Do we require this use case?

**Comment by jksuom** [comment]
It is my understanding that `evaluate=False` is mainly used for testing. Therefore this does not seem relevant for most uses of SymPy. But if some users find it useful, I think it could be included, possibly after some modifications.

**Comment by oscarbenjamin** [comment]
I think that it if evaluate=False actually worked in general then it could be used for lots of things. It needs a broader fix than this though (#17280). For example if `evaluate=False` then `Add/Mul.flatten` should not be called at all. A lot of code assumes that flattening has been done though and e.g. depends on the ordering of the args in an Add/Mul.

**Comment by cgabard** [comment]
For reference, if I remember correctly the full story, we used sympy in a teaching/test web application where student have to provide response to math exercice. We used sympy construction as base model of the mathematics expressions. We rely on it to reduce answer to compare with the solution expected. But in some use case, we did not want it. For example some exercice was to ask students to expand or factorize a specific mathematic expression. In this use case we did not want that sympy evaluate and change/simplify the expression. So having a correct behavior for this was important to avoid any modification.

**Comment by smichr** [comment]
For pedagogical purposes, symbol trickery (e.g. [here](https://stackoverflow.com/questions/53543562/sympy-treating-numbers-as-symbols/53544140#53544140)) can be used:
```python
>>> x + Symbol('0')
0 + x
>>> x*Symbol('1')
1*x
```

**Pull request closed without merging**